File: dynamic-programming/fibonacci/fibonacci.py
"""Dynamic Programming: Fibonacci with memoization"""

from functools import lru_cache


@lru_cache(maxsize=None)
def fib(n: int) -> int:
    if n <= 1:
        return n
    return fib(n - 1) + fib(n - 2)


if __name__ == "__main__":
    n = 20
    print(f"fib({n}) = {fib(n)}")

File: dynamic-programming/fibonacci/fibonacci.java
// Dynamic Programming: Fibonacci with memoization
import java.util.HashMap;
import java.util.Map;

class Fibonacci {

    static Map<Integer, Long> memo = new HashMap<>();

    static long fib(int n) {
        if (n <= 1) return n;
        if (memo.containsKey(n)) return memo.get(n);
        long result = fib(n - 1) + fib(n - 2);
        memo.put(n, result);
        return result;
    }

    public static void main(String[] args) {
        int n = 20;
        System.out.println("fib(" + n + ") = " + fib(n));
    }
}
